### PR TITLE
Doc: fix issue 1940 - ported Python gotchas from wiki to API docs

### PR DESCRIPTION
**Pull request by torimcd**
## What does this PR do?
Adds a new page to the API documentation that brings in the Python Gotchas from the wiki as described in https://github.com/OSGeo/gdal/issues/1940. Updates the API docs table of contents to link to the new page. 

## What are related issues/pull requests?
https://github.com/OSGeo/gdal/issues/1940
https://github.com/OSGeo/gdal/issues/1939

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

File: gdal/doc/source/api/index.rst [hunk 1]
```diff
@@ -82,6 +82,11 @@ API
    `Java API <../java/index.html>`_
    --------------------------------
 
+   .. toctree::
```
Comment (rouault) [inline gdal/doc/source/api/index.rst]: shouldn't that be moved right under the Python API paragraph ? In which case, we might want to remove the link from the Python API title itself, to have 2 entries, one that could be "API reference" and use the current <../python/index.html> and the other the python_gotchas . But not completely sure we can do that that way...

Comment (torimcd) [reply gdal/doc/source/api/index.rst]: Yes, I wasn't totally sure where the best spot for it would be but that makes sense. I'll take a look.

**Comment by rouault** [comment]
As I suspected Sphinx is not happy with the Link to an "external" html doc
```
/__w/1/s/gdal/doc/source/api/index.rst:82: WARNING: toctree contains reference to nonexisting document 'api/`API Reference <../python/index.html>`_'
```
I'm afraid we would have to make a hack like creating a target with empty content, that we would replace after build with a redirect to ../python/index.html . Probably out of scope of what you want to do... Maybe just revert to your initial proposition for now.

File: gdal/doc/source/api/python_gotchas.rst [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+================================================================================
+Python Gotchas in the GDAL and OGR Python Bindings
+================================================================================
+
+This page lists aspects of GDAL's and OGR's Python bindings that may catch Python programmers by surprise.
+If you find something new, feel free to add it to the list, but consider discussing it on the `gdal-dev mailing list <https://lists.osgeo.org/mailman/listinfo/gdal-dev>`__  first,
+to make sure you fully understand the issue and that others agree that it is unexpected, "non-Pythonic",
+or something that would catch many Python programmers by surprise.
+Be sure to reference email threads, Trac tickets, and other sources of additional information.
```
Comment (rouault) [inline gdal/doc/source/api/python_gotchas.rst]:
```suggestion
Be sure to reference email threads, GitHub tickets, and other sources of additional information.
```

File: gdal/doc/source/api/python_gotchas.rst [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+This list is not the place to report bugs. If you believe something is a bug, please `open a ticket <https://trac.osgeo.org/gdal/newticket>`__ and report the problem to gdal-dev.
```
Comment (rouault) [inline gdal/doc/source/api/python_gotchas.rst]:
```suggestion
This list is not the place to report bugs. If you believe something is a bug, please `open a ticket <https://github.com/OSGeo/gdal/issues>`__ and report the problem to gdal-dev.
```

File: gdal/doc/source/api/python_gotchas.rst [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+------------------------------------------------------------
+
+Python crashes in GDAL functions when you upgrade or downgrade numpy
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Much of GDAL's Python bindings are implemented in C++. Much of the core of numpy is implemented in C. The C++ part of GDAL's Python bindings interacts with the C part of numpy through numpy's ABI (application binary interface). 
+This requires GDAL's Python bindings to be compiled using numpy header files that define numpy C data structures. Those data structures sometimes change between numpy versions. When this happens, the new version of numpy is not be compatible at the binary level with the old version, and the GDAL Python bindings must be recompiled before they will work with the new verison of numpy. 
+And when they are recompiled, they probably won't work with the old version.
+
+If you obtained a precompiled version of GDAL's Python bindings, such as the Windows packages from `http://vbkto.dyndns.org/sdk/ <http://vbkto.dyndns.org/sdk/>`__ be sure you look up what version of numpy was used to compile them, and install that version of numpy on your machine.
```
Comment (rouault) [inline gdal/doc/source/api/python_gotchas.rst]:
```suggestion
If you obtained a precompiled version of GDAL's Python bindings, such as the Windows packages from `http://gisinternals.com/sdk.php <http://gisinternals.com/sdk.php>`__ be sure you look up what version of numpy was used to compile them, and install that version of numpy on your machine.
```

**Comment by rouault** [comment]
thanks!

**Comment by mloskot** [comment]
@torimcd Thanks!

**Pull request merged**